### PR TITLE
Getting ready to merge realm-java into master

### PR DESCRIPTION
**Pull request by kneth**
Instead of polluting `realm-java`, I have create a new branch. Almost except one unit test is working. The remaining failing test is cause by how core uses `mkdtemp` (see also https://code.google.com/p/android/issues/detail?id=74497) but I will make a workaround later. There are still a few issue with linking but I can link manually so I will fix that later too.

Please give me some feedback on the CMake parts of this PR - and only that.

Dependencies
 * https://github.com/realm/realm-core/pull/2355
 * https://github.com/realm/realm-sync/pull/1027

@cmelchior @emanuelez @bdash @radu-tutueanu

### CODE REVIEW

File: CMake/RealmCore.cmake [hunk 1]
```diff
@@ -238,25 +237,41 @@ endfunction()
...
-        ${USES_TERMINAL_BUILD}
-        ${ARGV}
-        )
+    if(REALM_PLATFORM STREQUAL "Android")
+      ExternalProject_Add(realm-core
+            PREFIX ${core_prefix_directory}
+            BUILD_IN_SOURCE 1
+            BUILD_COMMAND sh build.sh build-android
+            INSTALL_COMMAND ""
+            CONFIGURE_COMMAND REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS= sh build.sh config
```
Comment (bdash) [inline CMake/RealmCore.cmake]: The expectation when using existing local source trees for core and sync is that the trees are already configured.

File: CMake/RealmCore.cmake [hunk 2]
```diff
@@ -238,25 +237,41 @@ endfunction()
...
-        INSTALL_COMMAND ""
-        CONFIGURE_COMMAND ""
-        ${USES_TERMINAL_BUILD}
-        ${ARGV}
-        )
+    if(REALM_PLATFORM STREQUAL "Android")
+      ExternalProject_Add(realm-core
+            PREFIX ${core_prefix_directory}
+            BUILD_IN_SOURCE 1
+            BUILD_COMMAND sh build.sh build-android
```
Comment (bdash) [inline CMake/RealmCore.cmake]: Since this command should be the only difference, it'd be better to hoist it into a variable and leave only a single call to `ExternalProject_add`.

File: CMake/RealmCore.cmake [hunk 3]
```diff
@@ -238,25 +237,41 @@ endfunction()
...
+    if(REALM_PLATFORM STREQUAL "Android")
+        set(CONFIG_CMD REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS= sh build.sh config)
+    else()
+        set(CONFIG_CMD REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS=YES sh build.sh config)
+    endif()
     build_realm_core(GIT_REPOSITORY "https://github.com/realm/realm-core.git"
                      GIT_TAG ${branch}
-                     CONFIGURE_COMMAND test -f src/config.mk ||
-                                       REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS=YES sh build.sh config
+                     CONFIGURE_COMMAND test -f src/config.mk || CONFIG_CMD
```
Comment (bdash) [inline CMake/RealmCore.cmake]: This is currently breaking the build for this branch since you're not expanding the variable.

File: CMake/RealmCore.cmake [hunk 4]
```diff
@@ -272,22 +287,51 @@ endfunction()
...
-        ${USES_TERMINAL_BUILD}
-        ${ARGV}
-        )
+    if(REALM_PLATFORM STREQUAL "Android")
+        ExternalProject_Add(realm-sync-lib
+            DEPENDS realm-core
+            PREFIX ${cmake_files}/realm-sync
+            BUILD_IN_SOURCE 1
+            BUILD_COMMAND sh build.sh build-android
+            CONFIGURE_COMMAND rm -f src/config.mk && REALM_FORCE_OPENSSL=YES REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS= sh build.sh config && echo "ENABLE_ENCRYPTION    = yes" >> src/config.mk
```
Comment (bdash) [inline CMake/RealmCore.cmake]: Ditto about skipping configuration when using existing local source trees.

File: CMake/RealmCore.cmake [hunk 5]
```diff
@@ -272,22 +287,51 @@ endfunction()
...
-        INSTALL_COMMAND ""
-        ${USES_TERMINAL_BUILD}
-        ${ARGV}
-        )
+    if(REALM_PLATFORM STREQUAL "Android")
+        ExternalProject_Add(realm-sync-lib
+            DEPENDS realm-core
+            PREFIX ${cmake_files}/realm-sync
+            BUILD_IN_SOURCE 1
+            BUILD_COMMAND sh build.sh build-android
```
Comment (bdash) [inline CMake/RealmCore.cmake]: Ditto about hoisting the different command into a variable rather than repeating the entire `ExternalProject_Add`.

File: CMake/RealmCore.cmake [hunk 6]
```diff
@@ -342,10 +391,15 @@ endfunction()
 
 function(clone_and_build_realm_sync branch)
     set(cmake_files ${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY})
+    if(REALM_PLATFORM STREQUAL "Android")
+        set(CONFIG_CMD, rm -f src/config.mk && REALM_CORE_PREFIX=${cmake_files}/realm-core/src/realm-core REALM_FORCE_OPENSSL=YES REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS= sh build.sh config && echo "ENABLE_ENCRYPTION    = yes" >> src/config.mk)
```
Comment (bdash) [inline CMake/RealmCore.cmake]: The bits related to src/config.mk here are strange. Why does it always delete src/config.mk? Why does it explicitly add a line to src/config.mk after configuring?

File: CMake/RealmCore.cmake [hunk 7]
```diff
@@ -342,10 +391,15 @@ endfunction()
...
+    else()
+        set(CONFIG_CMD, test -f src/config.mk || REALM_CORE_PREFIX=${cmake_files}/realm-core/src/realm-core sh build.sh config)
```
Comment (bdash) [inline CMake/RealmCore.cmake]: The comma shouldn't be here.

File: CMake/RealmCore.cmake [hunk 8]
```diff
@@ -342,10 +391,15 @@ endfunction()
...
+    endif()
+
     build_realm_sync(GIT_REPOSITORY "git@github.com:realm/realm-sync.git"
                      GIT_TAG ${branch}
-                     CONFIGURE_COMMAND test -f src/config.mk ||
-                                       REALM_CORE_PREFIX=${cmake_files}/realm-core/src/realm-core sh build.sh config
+                     CONFIGURE_COMMAND CONFIG_CMD
```
Comment (bdash) [inline CMake/RealmCore.cmake]: You need to expand the variable.

File: src/impl/realm_coordinator.hpp [hunk 9]
```diff
@@ -145,7 +146,9 @@ class RealmCoordinator : public std::enable_shared_from_this<RealmCoordinator> {
     std::unique_ptr<_impl::ExternalCommitHelper> m_notifier;
     std::function<void(VersionID, VersionID)> m_transaction_callback;
 
+#if REALM_ENABLE_SYNC
```
Comment (bdash) [inline src/impl/realm_coordinator.hpp]: Doesn't appear to be related to the Java build.

File: src/shared_realm.hpp [hunk 10]
```diff
@@ -31,6 +31,8 @@
 #include <memory>
 #include <thread>
 
+#include <realm/group_shared.hpp>
```
Comment (bdash) [inline src/shared_realm.hpp]: I provided feedback about this, and several other parts of the code in this PR, on one of the other Java PRs. Can you please take a look at the previous PRs so I don't have to repeat it here?

**Comment by beeender** [comment]
@kneth Is the cmake changes necessary? I think we'd better only have minimal changes in this PR. BTW, I will fix the `shared_group.h` header problem and push to this PR.

File: CMake/RealmCore.cmake (hunk 6 above)
Comment (bdash) [inline CMake/RealmCore.cmake]: I can't see how adding `ENABLE_ENCRYPTION` to `src/config.mk` could possibly be required. No variable of that name is ever referenced in the realm-sync codebase. Whether or not encryption is enabled is determined solely by core's configuration.

File: CMake/RealmCore.cmake [hunk 11]
```diff
@@ -272,22 +281,44 @@ endfunction()
...
+        endif()
+    endif()
+
+    
     ExternalProject_Get_Property(realm-sync-lib SOURCE_DIR)
     set(sync_directory ${SOURCE_DIR})
 
-    set(sync_library_debug ${sync_directory}/src/realm/librealm-sync-dbg.a)
-    set(sync_library_release ${sync_directory}/src/realm/librealm-sync.a)
+    if(REALM_PLATFORM STREQUAL "Android")
```
Comment (bdash) [inline CMake/RealmCore.cmake]: The path to the built libraries could also be hoisted into the Android-vs-other if block above, so we don't have to duplicate the list of libraries for both sync and the sync server.

File: CMake/RealmCore.cmake [hunk 12]
```diff
@@ -342,10 +378,15 @@ endfunction()
 
 function(clone_and_build_realm_sync branch)
     set(cmake_files ${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY})
+    if(REALM_PLATFORM STREQUAL "Android")
+        set(config_cmd test -f src/config.mk || REALM_CORE_PREFIX=${cmake_files}/realm-core/src/realm-core REALM_FORCE_OPENSSL=YES REALM_ENABLE_ENCRYPTION=YES REALM_ENABLE_ASSERTIONS= sh build.sh config && echo "ENABLE_ENCRYPTION    = yes" >> src/config.mk)
```
Comment (bdash) [inline CMake/RealmCore.cmake]: As stated previously, there's no need to set `REALM_ENABLE_ENCRYPTION` here, or to add `ENABLE_ENCRYPTION` to src/config.mk. realm-sync doesn't use these.

File: src/shared_realm.cpp [hunk 13]
```diff
@@ -364,10 +364,13 @@ void Realm::update_schema(Schema schema, uint64_t version, MigrationFunction mig
...
 
     m_group->set_schema_change_notification_handler(nullptr);
-    transaction::begin_without_validation(*m_shared_group);
+    if (!in_transaction) {
+        transaction::begin_without_validation(*m_shared_group);
+    }
     add_schema_change_handler();
 
     // Cancel the write transaction if we exit this function before committing it
+    // FIXME: don't cancel if in_transaction == true
```
Comment (bdash) [inline src/shared_realm.cpp]: Why add this FIXME rather than adding the `!in_transaction` check to `WriteTransactionGuard`'s destructor? Won't the behavior as-is result in some subtle bugs?

File: src/shared_realm.hpp [hunk 14]
```diff
@@ -168,6 +169,9 @@ class Realm : public std::enable_shared_from_this<Realm> {
 
         /// A data structure storing data used to configure the Realm for sync support.
         std::shared_ptr<SyncConfig> sync_config;
+
+        // FIXME: realm java is manage sync in the java level, but needs Realm created sync history.
```
Comment (bdash) [inline src/shared_realm.hpp]: The wording here is awkward. Perhaps something like:

> Realm Java manages sync at the Java level, so it needs to create Realms using the sync history format.

File: tests/sync/sync_test_utils.cpp [hunk 15]
```diff
@@ -49,7 +49,11 @@ std::string tmp_dir() {
     const char* dir = getenv("TMPDIR");
     if (dir && *dir)
         return dir;
-    return "/tmp";
+#if REALM_ANDROID
+        return "/data/local/tmp";
```
Comment (bdash) [inline tests/sync/sync_test_utils.cpp]: Indentation here is strange.

File: CMake/RealmCore.cmake (hunk 12 above)
Comment (kneth) [inline CMake/RealmCore.cmake]: If `ENABLE_ENCRYPTION` isn't set, OpenSSL will _not_ be downloaded, built and included. `REALM_ENABLE_ENCRYPTION=YES` is not required (it is ignored by realm-sync's `build.sh`).

Comment (bdash) [inline CMake/RealmCore.cmake]: Can you clarify what uses `ENABLE_ENCRYPTION`? I see nothing in realm-sync that does. AFAICT `build.sh build-android` unconditionally [downloads](https://github.com/realm/realm-sync/blob/7b311c1bc19388608e52e984f640ed7f5c98b114/build.sh#L808) and [builds](https://github.com/realm/realm-sync/blob/7b311c1bc19388608e52e984f640ed7f5c98b114/build.sh#L872-L899) OpenSSL. Are you perhaps thinking of realm-core, which does use `ENABLE_ENCRYPTION`?
